addon/components/frost-popover.js

```diff
@@ -1,96 +1,328 @@
 import Ember from 'ember'
 import layout from '../templates/components/frost-popover'
 import $ from 'jquery'
-export default Ember.Component.extend({
+import PropTypeMixin, {PropTypes} from 'ember-prop-types'
+import {checkLeft, checkRight, checkTop, checkBottom} from './util'
+
+const arrowMargin = 5
+const maxPlacementRetries = 5
+
+export default Ember.Component.extend(PropTypeMixin, {
   layout,
-  visible: false,
-  event: 'click',
-  closest: false,
-  offset: 10,
-  position: 'bottom',
-  index: 0,
-  excludePadding: false,
-  classNameBindings: ['position', 'visible:visible:invisible'],
+  classNameBindings: ['visible:visible:invisible', 'autoPosition'],
   classNames: ['tooltip-frost-popover'],
-  didRender () {
+  propTypes: {
+    closest: PropTypes.bool,
+    event: PropTypes.string,
+    excludePadding: PropTypes.bool,
+    index: PropTypes.number,
+    offset: PropTypes.number,
+    position: PropTypes.string,
+    resize: PropTypes.bool,
+    viewport: PropTypes.oneOfType([
+      PropTypes.string,
+      PropTypes.object,
+      PropTypes.func
+    ])
+  },
+
+  getDefaultProps () {
+    return {
+      closest: false,
+      event: 'click',
+      excludePadding: false,
+      index: 0,
+      offset: 10,
+      position: 'bottom',
+      resize: true,
+      viewport: 'body',
```

Comment:
Dude, I would love to see this used by `frost-select`. It'll get it for free if it did.